File: .shipmate/hooks/completion-reflection.py
#!/usr/bin/env python3
"""
Completion Reflection Hook (stop)

Cursor Hook: Performs brief self-reflection when agent completes a task.
Logs successful patterns and learnings for compounding engineering.

Input (stdin JSON):
{
  "conversation_id": "string",
  "generation_id": "string",
  "model": "string",
  "hook_event_name": "stop",
  "cursor_version": "string",
  "workspace_roots": ["path"],
  "user_email": "string | null",
  "status": "string",
  "loop_count": number
}

Output (stdout JSON):
{
  "followup_message": "optional message for agent to process"
}
"""

import json
import sys
import os
import uuid
from datetime import datetime
from pathlib import Path

# Track conversation stats in memory (reset per hook invocation)
# For persistent tracking, we'd need to read from diary entries
CONVERSATION_STATS_FILE = Path.home() / ".shipmate" / "diary" / ".conversation_stats.json"


def get_diary_path() -> Path:
    """Get the diary directory path, creating if needed."""
    diary_dir = Path.home() / ".shipmate" / "diary" / "entries" / datetime.now().strftime("%Y-%m-%d")
    diary_dir.mkdir(parents=True, exist_ok=True)
    return diary_dir


def get_conversation_stats(conversation_id: str) -> dict:
    """Get stats for this conversation from today's diary entries."""
    stats = {
        "corrections": 0,
        "errors": 0,
        "files_edited": [],
        "commands_run": 0
    }

    diary_dir = Path.home() / ".shipmate" / "diary" / "entries" / datetime.now().strftime("%Y-%m-%d")
    if not diary_dir.exists():
        return stats

    for entry_file in diary_dir.glob("*.json"):
        try:
            with open(entry_file, 'r') as f:
                entry = json.load(f)

            if entry.get("conversation_id") != conversation_id:
                continue

            entry_type = entry.get("type", "")
            if entry_type == "correction":
                stats["corrections"] += 1
            elif entry_type == "error_recovery":
                stats["errors"] += 1
            elif entry_type == "shell_execution":
                stats["commands_run"] += 1
                # Track file edits from command context if available
            elif entry_type == "file_edit":
                file_path = entry.get("context", {}).get("file_path", "")
                if file_path and file_path not in stats["files_edited"]:
                    stats["files_edited"].append(file_path)

        except Exception:
            continue

    return stats


def calculate_session_quality(stats: dict, loop_count: int) -> dict:
    """Calculate quality metrics for the session."""
    quality = {
        "score": 0.5,  # Base score
        "factors": []
    }

    # Corrections reduce quality
    if stats["corrections"] > 0:
        penalty = min(stats["corrections"] * 0.1, 0.3)
        quality["score"] -= penalty
        quality["factors"].append(f"-{penalty:.1f} for {stats['corrections']} correction(s)")

    # Errors reduce quality (but less than corrections - errors are normal)
    if stats["errors"] > 2:
        penalty = min((stats["errors"] - 2) * 0.05, 0.15)
        quality["score"] -= penalty
        quality["factors"].append(f"-{penalty:.1f} for {stats['errors']} errors")

    # File edits are positive signals
    if len(stats["files_edited"]) > 0:
        bonus = min(len(stats["files_edited"]) * 0.05, 0.2)
        quality["score"] += bonus
        quality["factors"].append(f"+{bonus:.1f} for {len(stats['files_edited'])} file(s) edited")

    # Completing with low loop count is positive
    if loop_count <= 3:
        quality["score"] += 0.1
        quality["factors"].append("+0.1 for efficient completion")

    quality["score"] = max(0.0, min(1.0, quality["score"]))
    return quality


def create_completion_entry(input_data: dict, stats: dict, quality: dict) -> dict:
    """Create a diary entry for the completion."""
    entry = {
        "id": str(uuid.uuid4()),
        "timestamp": datetime.now().isoformat(),
        "conversation_id": input_data.get("conversation_id", "unknown"),
        "type": "completion",
        "severity": "low",
        "category": "workflow",
        "context": {
            "status": input_data.get("status", "unknown"),
            "loop_count": input_data.get("loop_count", 0),
            "model": input_data.get("model", "unknown"),
            "workspace_roots": input_data.get("workspace_roots", []),
            "session_stats": stats
        },
        "learning": {
            "what_happened": f"Session completed with status: {input_data.get('status', 'unknown')}",
            "quality_score": quality["score"],
            "quality_factors": quality["factors"],
            "corrections_count": stats["corrections"],
            "errors_count": stats["errors"],
            "files_edited_count": len(stats["files_edited"]),
            "improvement": "Reinforce successful patterns" if quality["score"] >= 0.6 else "Analyze for improvements"
        },
        "metadata": {
            "hook_version": "1.0.0",
            "cursor_version": input_data.get("cursor_version", "unknown")
        }
    }

    return entry


def save_diary_entry(entry: dict) -> str:
    """Save diary entry to file."""
    diary_path = get_diary_path()
    filename = f"completion-{entry['id'][:8]}.json"
    filepath = diary_path / filename

    with open(filepath, 'w') as f:
        json.dump(entry, f, indent=2)

    return str(filepath)


def main():
    try:
        input_data = json.load(sys.stdin)
    except json.JSONDecodeError:
        print(json.dumps({}))
        sys.exit(0)
    except Exception:
        print(json.dumps({}))
        sys.exit(0)

    conversation_id = input_data.get("conversation_id", "")
    loop_count = input_data.get("loop_count", 0)

    # Get conversation stats from today's diary
    stats = get_conversation_stats(conversation_id)

    # Calculate quality metrics
    quality = calculate_session_quality(stats, loop_count)

    # Create and save completion entry
    entry = create_completion_entry(input_data, stats, quality)
    save_diary_entry(entry)

    # Optionally provide a followup message for low-quality sessions
    output = {}
    if quality["score"] < 0.4 and stats["corrections"] > 2:
        output["followup_message"] = "Note: This session had multiple corrections. Consider reviewing the approach."

    print(json.dumps(output))
    sys.exit(0)


if __name__ == "__main__":
    main()

File: .shipmate/hooks/ts/completion-reflection.ts
#!/usr/bin/env bun

/**
 * Completion Reflection Hook (stop)
 *
 * Brief self-reflection when agent stops/completes.
 */

import { createHash } from 'node:crypto';
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { homedir } from 'node:os';
import { join } from 'node:path';

interface HookInput {
  conversation_id: string;
  hook_event_name: string;
  status?: string;
  loop_count?: number;
}

interface HookOutput {
  followup_message?: string;
}

interface SessionStats {
  files_edited: string[];
  total_lines_added: number;
  total_lines_removed: number;
  file_types: Record<string, number>;
  edit_count: number;
  last_update: number;
}

const STATS_FILE = join(homedir(), '.shipmate', 'context', 'session_stats.json');
const CORRECTIONS_FILE = join(homedir(), '.shipmate', 'context', 'active_corrections.json');

function loadSessionStats(conversationId: string): SessionStats {
  const empty: SessionStats = {
    files_edited: [],
    total_lines_added: 0,
    total_lines_removed: 0,
    file_types: {},
    edit_count: 0,
    last_update: 0,
  };

  if (!existsSync(STATS_FILE)) return empty;

  try {
    const allStats = JSON.parse(readFileSync(STATS_FILE, 'utf-8'));
    return allStats[conversationId] || empty;
  } catch {
    return empty;
  }
}

function countCorrections(conversationId: string): number {
  if (!existsSync(CORRECTIONS_FILE)) return 0;

  try {
    const corrections = JSON.parse(readFileSync(CORRECTIONS_FILE, 'utf-8'));
    const conv = corrections[conversationId];
    return conv ? conv.detections?.length || 0 : 0;
  } catch {
    return 0;
  }
}

interface QualityMetrics {
  score: number;
  factors: string[];
}

function calculateSessionQuality(
  stats: SessionStats,
  corrections: number,
  loopCount: number,
): QualityMetrics {
  const quality: QualityMetrics = { score: 0.5, factors: [] };

  // Corrections reduce quality
  if (corrections > 0) {
    const penalty = Math.min(corrections * 0.1, 0.3);
    quality.score -= penalty;
    quality.factors.push(`${corrections} corrections (-${(penalty * 100).toFixed(0)}%)`);
  }

  // Files edited increase quality
  if (stats.files_edited.length > 0) {
    const bonus = Math.min(stats.files_edited.length * 0.05, 0.2);
    quality.score += bonus;
    quality.factors.push(
      `${stats.files_edited.length} files edited (+${(bonus * 100).toFixed(0)}%)`,
    );
  }

  // High loop counts might indicate issues
  if (loopCount > 10) {
    const penalty = Math.min((loopCount - 10) * 0.02, 0.15);
    quality.score -= penalty;
    quality.factors.push(`High loop count: ${loopCount} (-${(penalty * 100).toFixed(0)}%)`);
  }

  // Net lines added is a positive signal
  const netLines = stats.total_lines_added - stats.total_lines_removed;
  if (netLines > 0) {
    const bonus = Math.min(netLines * 0.001, 0.1);
    quality.score += bonus;
    quality.factors.push(`+${netLines} net lines (+${(bonus * 100).toFixed(0)}%)`);
  }

  quality.score = Math.max(0, Math.min(1, quality.score));
  return quality;
}

function logCompletion(
  input: HookInput,
  stats: SessionStats,
  corrections: number,
  quality: QualityMetrics,
) {
  const now = new Date();
  const dateStr = now.toISOString().split('T')[0];
  const diaryDir = join(homedir(), '.shipmate', 'diary', 'entries', dateStr);

  try {
    mkdirSync(diaryDir, { recursive: true });

    const id = createHash('md5')
      .update(`${input.conversation_id}${now.toISOString()}`)
      .digest('hex')
      .slice(0, 8);

    const entry = {
      id,
      timestamp: now.toISOString(),
      conversation_id: input.conversation_id,
      type: 'completion',
      severity: quality.score < 0.4 ? 'high' : quality.score < 0.6 ? 'medium' : 'low',
      category: 'workflow',
      context: {
        status: input.status,
        loop_count: input.loop_count,
        files_edited: stats.files_edited,
        lines_added: stats.total_lines_added,
        lines_removed: stats.total_lines_removed,
        correction_count: corrections,
      },
      learning: {
        what_happened: 'Session completed',
        quality_score: quality.score,
        quality_factors: quality.factors,
        improvement:
          quality.score < 0.6 ? 'Review session for improvements' : 'Patterns working well',
      },
      metadata: {
        hook_version: '1.0.0',
      },
    };

    writeFileSync(join(diaryDir, `completion-${id}.json`), JSON.stringify(entry, null, 2));
  } catch {
    // Silently fail
  }
}

function cleanupSessionData(conversationId: string) {
  // Clean up session stats
  if (existsSync(STATS_FILE)) {
    try {
      const allStats = JSON.parse(readFileSync(STATS_FILE, 'utf-8'));
      if (conversationId in allStats) {
        delete allStats[conversationId];
        writeFileSync(STATS_FILE, JSON.stringify(allStats, null, 2));
      }
    } catch {
      // Silently fail
    }
  }

  // Clean up corrections
  if (existsSync(CORRECTIONS_FILE)) {
    try {
      const corrections = JSON.parse(readFileSync(CORRECTIONS_FILE, 'utf-8'));
      if (conversationId in corrections) {
        delete corrections[conversationId];
        writeFileSync(CORRECTIONS_FILE, JSON.stringify(corrections, null, 2));
      }
    } catch {
      // Silently fail
    }
  }
}

async function main() {
  try {
    const input: HookInput = await Bun.stdin.json();
    const output: HookOutput = {};

    const stats = loadSessionStats(input.conversation_id);
    const corrections = countCorrections(input.conversation_id);
    const quality = calculateSessionQuality(stats, corrections, input.loop_count || 0);

    // Log completion
    logCompletion(input, stats, corrections, quality);

    // Clean up session data
    cleanupSessionData(input.conversation_id);

    // Optionally provide feedback for poor quality sessions
    if (quality.score < 0.4) {
      output.followup_message = `Session quality was low (${(quality.score * 100).toFixed(0)}%). Consider reviewing the diary for improvements.`;
    }

    console.log(JSON.stringify(output));
  } catch {
    console.log(JSON.stringify({}));
  }
}

main();
